### PR TITLE
ECDC-2019 use p4p for PV Access

### PR DESCRIPTION
**Pull request by mattclarke**
This implements support for using PVA via the p4p library.
The changes also mean the NICOS will use EPICS monitors rather than polling

### CODE REVIEW

**Comment by ebadkamil** [comment]
With this PR, do you also want to get rid of `nicos_ess.devices.epics.base` and `nicos_ess.devices.epics.motor` ?  Then it will be great if we can remove those classes in this particular PR itself and any dependence on those classes should be appropriately replaced with the new implementations.

File: nicos_ess/devices/epics/pva/epics_devices.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+        self._pvs = {}
+
+        if mode != SIMULATION:
+            for pvparam in self._get_pv_parameters():
+                # Retrieve the actual PV name
+                pvname = self._get_pv_name(pvparam)
+                if not pvname:
+                    raise ConfigurationError(self, 'PV for parameter %s was '
+                                                   'not found!' % pvparam)
+                # Check pv exists - throws if cannot connect
```
Comment (ebadkamil) [inline nicos_ess/devices/epics/pva/epics_devices.py:169]:
```suggestion
                # Check pv exists - throws CommunicationError if cannot connect
```

File: nicos_ess/devices/epics/pva/p4p.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+import numpy as np
+from p4p.client.thread import Context
+
+from nicos.core import CommunicationError, status
+
+# Same context can be shared across all devices.
+# nt=False tells p4p not to try to map types itself
+# we want to do this manually to avoid information loss
+from nicos.devices.epics import SEVERITY_TO_STATUS
```
Comment (ebadkamil) [inline nicos_ess/devices/epics/pva/p4p.py:35]: Move this import above the comments for context.

**Comment by mattclarke** [comment]
Great question.
No, we (unfortunately) need to still be able to use those as not every device will have PVA support initially.
Also, SINQ are reliant on the CA versions.

Once the CA monitor support is in (https://forge.frm2.tum.de/review/c/frm2/nicos/nicos/+/22768) we can see if we can extract the common functionality out so there is less duplication.

File: nicos_ess/devices/epics/pva/p4p.py (hunk 2 above)
Comment (mattclarke) [reply nicos_ess/devices/epics/pva/p4p.py:35]: 👍

File: nicos_ess/devices/epics/pva/epics_devices.py (hunk 1 above)
Comment (mattclarke) [reply nicos_ess/devices/epics/pva/epics_devices.py:169]: Not a good idea as it couples the comment to code in a different file.
If someone changes the Exception type later then you are relying on them to also to remember to update the comment.

File: nicos_ess/devices/epics/pva/p4p.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    def __init__(self):
+        self.disconnected = set()
+        self.lock = RLock()
+
+    def connect_pv(self, pvname, timeout):
+        # Check pv is available
+        try:
+            _CONTEXT.get(pvname, timeout=timeout)
+        except TimeoutError:
+            raise CommunicationError('could not connect to PV %r' % pvname)
```
Comment (ebadkamil) [inline nicos_ess/devices/epics/pva/p4p.py:49]: Just for curiosity
Is pvname a string like "SIMPLE:VALUE1" or is it of type Param ? If it's former, can we use f strings here
```python
raise CommunicationError(f'could not connect to PV {pvname}')
```

Comment (mattclarke) [reply nicos_ess/devices/epics/pva/p4p.py:49]: No, we should definitely do that - thanks!

File: nicos_ess/devices/epics/pva/p4p.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        result = _CONTEXT.get(pv, timeout=timeout)
+        try:
+            default_low = result['display']['limitLow']
+            default_high = result['display']['limitHigh']
+        except KeyError:
+            pass
+        return default_low, default_high
+
+    def subscribe(self, pvname, pvparam, change_callback,
+                  connection_callback=None, as_string=False):
```
Comment (ebadkamil) [inline nicos_ess/devices/epics/pva/p4p.py:134]: Docstring will be very helpful for this function.

File: nicos_ess/devices/epics/pva/epics_devices.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+#
+# *****************************************************************************
+
+"""
+This module contains some classes for NICOS - EPICS integration.
+"""
+import time
+
+from nicos import session
+from nicos.core import POLLER, SIMULATION, ConfigurationError,\
```
Comment (ebadkamil) [inline nicos_ess/devices/epics/pva/epics_devices.py:31]:
```suggestion
from nicos.core import (
    anytype, ConfigurationError, DeviceMixinBase, 
    floatrange, HasLimits, Moveable, none_or, 
    Override, Param, POLLER, pvname, Readable, 
    SIMULATION, status)
```

Comment (mattclarke) [reply nicos_ess/devices/epics/pva/epics_devices.py:31]: No, it has to be done the NICOS way.

Comment (ebadkamil) [reply nicos_ess/devices/epics/pva/epics_devices.py:31]: also no policy on alphabetical imports in Nicos?

Comment (mattclarke) [reply nicos_ess/devices/epics/pva/epics_devices.py:31]: isort https://pypi.org/project/isort/

Comment (jkrueger1) [reply nicos_ess/devices/epics/pva/epics_devices.py:31]: The order is in ASCII order: A-Za-z, which means that all entries with upper case letter first. As @mattclarke mentioned we use the isort tool and the import is not encapsulated by '()'.

Comment (ebadkamil) [reply nicos_ess/devices/epics/pva/epics_devices.py:31]: super. Thank you both.

**Pull request merged**